conda/gateways/connection/download.py

```diff
@@ -209,6 +118,126 @@ def download(
         )
 
 
+@time_recorder("download")
+def download(
+    url,
+    target_full_path,
+    md5=None,
+    sha256=None,
+    size=None,
+    progress_update_callback=None,
+):
+    if exists(target_full_path):
+        maybe_raise(BasicClobberError(target_full_path, url, context), context)
+    if not context.ssl_verify:
+        disable_ssl_verify_warning()
+
+    with download_http_errors(url):
+        return _download_inner(
+            url, target_full_path, md5, sha256, size, progress_update_callback
+        )
+
+
+def _download_inner(url, target_full_path, md5, sha256, size, progress_update_callback):
+    timeout = context.remote_connect_timeout_secs, context.remote_read_timeout_secs
+    session = get_session(url)
+    resp = session.get(url, stream=True, proxies=session.proxies, timeout=timeout)
+    if log.isEnabledFor(DEBUG):
+        log.debug(stringify(resp, content_max_len=256))
+    resp.raise_for_status()
+
+    content_length = int(resp.headers.get("Content-Length", 0))
+
+    # prefer sha256 over md5 when both are available
+    checksum_type = checksum = None
+    if sha256:
+        checksum_builder = hashlib.new("sha256")
+        checksum_type = "sha256"
+        checksum = sha256
+    elif md5:
+        checksum_builder = hashlib.new("md5")
+        checksum_type = "md5"
+        checksum = md5
+    else:
+
+        class NullChecksumBuilder:
+            """
+            Simplify checksum-update code when no checksum.
+            """
+
+            def update(self, bytes):
+                return
+
+            def hexdigest(self):
+                raise NotImplemented()
+
+        checksum_builder = NullChecksumBuilder()
+
+    size_builder = 0
+
+    with open(target_full_path, "wb") as fh:
+        streamed_bytes = 0
+        for chunk in resp.iter_content(chunk_size=CHUNK_SIZE):
+            # chunk could be the decompressed form of the real data
+            # but we want the exact number of bytes read till now
+            streamed_bytes = resp.raw.tell()
+            try:
+                fh.write(chunk)
+            except OSError as e:
+                message = "Failed to write to %(target_path)s\n  errno: %(errno)d"
+                # TODO: make this CondaIOError
+                raise CondaError(message, target_path=target_full_path, errno=e.errno)
+
+            checksum_builder.update(chunk)
```

Comment:
updated version should checksum-on-file at the end since we can't resume a checksum